### ISSUE TITLE
Assisted Inject with new string qualifiers

### ISSUE THREAD

**Issue by z3ntu** [google/dagger#2373]
Since the latest version 2.32 something like the following code isn't possible to use anymore because of the change that the type and identifier combo must be unique:
```kotlin
interface FoobarFactory<T : ViewModel> {
    fun create(environmentId: String, wizardId: String, stepId: String): T
}

// there are multiple of such viewmodels that all use those three assisted string parameters,
// which is why FoobarFactory exists.
class FooViewModel @AssistedInject constructor(
    @Assisted environmentId: String,
    @Assisted wizardId: String,
    @Assisted stepId: String
) : ... {
    @AssistedFactory
    interface Factory : FoobarFactory<FooViewModel>
}
```

I did try several ways to make this code compatible with 2.32 by e.g. changing `@Assisted` in the viewmodel to `@Assisted("parametername")` and also adding `@Assisted("parametername")` to the 'template' interface (`FoobarFactory`) but essentially just this worked (in addition to adding the qualifier to the viewmodel argument of course):
```kotlin
class FooViewModel ... {
    @AssistedFactory
    interface Factory : FoobarFactory<FooViewModel> {
        override fun create(
            @Assisted("environmentId") environmentId: String,
            @Assisted("wizardId") wizardId: String,
            @Assisted("stepId") stepId: String
        ): FooViewModel
    }
}
```
As I have multiple of these FooViewModel factories I'd prefer to not have to duplicate and override the function several times.

**Comment by FunkyMuse** [comment]
Yup the error throw is

``kotlin
error: @AssistedInject constructor has duplicate @Assisted type: @Assisted java.lang.String
``
Pretty annoying

**Comment by FunkyMuse** [comment]
@Chang-Eric I don't think this is a good idea, for example if I have 

```kotlin
class SubtitlesVM @AssistedInject constructor(
        @Assisted private val language: String,
        @Assisted private val languageFullName: String,
) : ViewModel() {


    @AssistedFactory
    interface SubtitlesVMFactory {
        fun create(language: String, languageFullName: String): SubtitlesVM
    }
}
```
Then the parameter of the view model name matches the one in the factory, why would i need it named?
Perhaps if the name doesn't match then we can not use the @Assisted(name = "someNameThatMatches").

This change doesn't make any sense and it's a choice that shouldn't have been made.

**Comment by danysantiago** [comment]
@FunkyMuse, parameter name matching is not reliable when processing already compiled classes, say from another library or due to incremental compilation, it is also not included by default in the bytecode. See the discussion in https://github.com/google/dagger/issues/2281

**Comment by FunkyMuse** [comment]
> @FunkyMuse, parameter name matching is not reliable when processing already compiled classes, say from another library or due to incremental compilation, it is also not included by default in the bytecode. See the discussion in #2281

I think Named would be more convenient instead of just doubling the assisted naming 
```kotlin
class SubtitlesVM @AssistedInject constructor(
        @Assisted private val language: String,
        @Assisted private val languageFullName: String,
) : ViewModel() {


    @AssistedFactory
    interface SubtitlesVMFactory {
        fun create(@Named("language") language: String, @Named("languageFullName") languageFullName: String): SubtitlesVM
    }
}
```

**Comment by z3ntu** [comment]
Just to be clear I don't mind having to provide the names like `@Assisted("foobar")`, it's just that I can't provide that annotation to the parameter in `FoobarFactory` as that isn't a `@AssistedFactory` (and can't be because of the generics)

**Comment by Chang-Eric** [comment]
@FunkyMuse I haven't fully thought through the cases, but it is possible you might need `@Named` on both sites, since there may be cases where we are missing the parameter names from the `@AssistedInject` class too. There's also a question of how intuitive it is that in one class the parameter names matter and are the standard (in the `@AssistedInject` class) and in one class (the factory) the parameter names are essentially ignored.

I'm also not sure I understand the issue with the default name being the empty string? The matching is done via a combo of the string + type, so it is essentially the same as qualifier annotations, where "" is the same as no qualifier annotation. Yes, people could do `@Assisted("")` and make it look funny, but seems like a rare case.

@z3ntu I assume you're hitting one of our checks on usage of `@Assisted`? If so, we may need to look at relaxing that constraint.

**Comment by z3ntu** [comment]
So this didn't work:
```kotlin
interface FoobarFactory<T : ViewModel> {
    fun create(@Assisted("environmentId") environmentId: String, @Assisted("wizardId") wizardId: String, @Assisted("stepId") stepId: String): T
}
```
and then inheriting from that interface like this without declaring the function there
```kotlin
    @AssistedFactory
    interface Factory : FoobarFactory<FooViewModel>
```

**Comment by FunkyMuse** [comment]
> @FunkyMuse I haven't fully thought through the cases, but it is possible you might need `@Named` on both sites, since there may be cases where we are missing the parameter names from the `@AssistedInject` class too. There's also a question of how intuitive it is that in one class the parameter names matter and are the standard (in the `@AssistedInject` class) and in one class (the factory) the parameter names are essentially ignored.
> 
> I'm also not sure I understand the issue with the default name being the empty string? The matching is done via a combo of the string + type, so it is essentially the same as qualifier annotations, where "" is the same as no qualifier annotation. Yes, people could do `@Assisted("")` and make it look funny, but seems like a rare case.
> 
> @z3ntu I assume you're hitting one of our checks on usage of `@Assisted`? Just want to confirm since in your first example `FoobarFactory` is marked with `@AssistedFactory`. If so, we may need to look at relaxing that constraint.

The empty string is a default value, so if we want to enforce named parameters we need a required value

Assisted implementation is 

`@Assisted(val name : String = "")`

One can get really confusef about the usability of the name since by default is an empty string and it doesn't enforce you to name the parameter, that's why qualifiers are the best in this case or something like

`@AssistedParameter(val name : String)`

Where the compiler will require the user to input a name, but introducing another annotation, I don't know, seems a bit overkill where you can achieve something like this with qualifiers.

**Comment by Chang-Eric** [comment]
@z3ntu Thanks for that example. I agree that is not ideal. WDYT about if we added something to suppress that check? We still think it is valuable to check for people mistakenly adding `@Assisted` to other classes where it won't be used. It would require another annotation on your interface like `@DisableAssistedCheck` or something, but hopefully that isn't too much trouble and the cases where you have to disambiguate parameters and use an interface may be rarer?

@FunkyMuse I think maybe there is some confusion here. We actually don't want to generally enforce using a name parameter. Only in the cases where the type of the arguments make it ambiguous. That's why empty string is the default, because if you have a factory with just `Foo` and `Bar` arguments, we don't want to force you to additionally name them.

**Comment by Chang-Eric** [comment]
@z3ntu Btw, I think a workaround could be redeclaring/overriding the method in the `@AssistedFactory` with the `@Assisted` annotations. For example:

```
interface FoobarFactory<T : ViewModel> {
    fun create(environmentId: String, wizardId: String, stepId: String): T
}

@AssistedFactory
interface Factory : FoobarFactory<FooViewModel> {
  override fun create(@Assisted("environmentId") environmentId: String, @Assisted("wizardId") wizardId: String, @Assisted("stepId") stepId: String): T
}
```

**Comment by z3ntu** [comment]
@Chang-Eric Yes your codeblock matches the second codeblock in my original comment, it's just less than ideal for me having to copy the override multiple times to the different interfaces.

**Comment by Chang-Eric** [comment]
@z3ntu Oops, sorry I missed that, getting confused with the conversations. Did you have an opinion on a suppression annotation?

**Comment by z3ntu** [comment]
Should be fine. It's quite a niche use case that I have, so explicitly saying "yes I want to really do this" is probably a nice option.
